### ISSUE TITLE
Alternative to SHACL validation requiring inferencing

### ISSUE THREAD

**Issue by edmondchuc** [ternaustralia/ontology_tern#95]
At the moment, the target instances must contain the target class to pass validation. 

Example, an observation's `sosa:hasFeatureOfInterest` must have a value which states it is a `tern:FeatureOfInterest` to pass validation even if the target data states it is one of the specialised classes, e.g., `tern:MaterialSample`. 

The current solution is to either explicitly state the required class in the data or enable inferencing during validation.

What I propose is to run the SHACL shapes definitions through a processor which expands the target classes to include all of the specialised classes of the original target class. This will ensure the data is conformant without downstream users requiring inferencing or the addition of explicit statements to validate their data.

**Comment by edmondchuc** [comment]
On second thought, this won't cover the specialised classes "inheriting" the constraints of its superclasses.

For example, a [Site](https://linkeddata.tern.org.au/viewers/tern-ontology?uri=https://w3id.org/tern/ontologies/tern/Site) must follow all the constraints of its superclasses, and that's only possible if the data explicitly states a Site is also an instance of its superclasses or the triples are materialised with a reasoner.

**Comment by dr-shorthair** [comment]
> run the SHACL shapes definitions through a processor which expands the target classes to include all of the specialised classes of the original target class

Do you mean to add all of the superclasses into the data, or all of the subclasses into the validation scripts?

**Comment by edmondchuc** [comment]
> Do you mean to add all of the superclasses into the data, or all of the subclasses into the validation scripts?

Yes. The former requires users to expand their data graph based on the TERN Ontology with an RDFS reasoner to use the current SHACL validation.

I was trying to see if there is a way to programatically expand what the SHACL shapes are targeting. For example,  `tern:Result` is a superclass to `Sample`, `MaterialSample`, `Quadrat`, `Site` and `Transect`. The current shape for `Result` only targets `tern:Result`, but we can expand it programatically to also target all of its subclasses. I think this should work, I have to test it.
